internal/core/src/query/GroupByOperator.cpp

```diff
@@ -0,0 +1,235 @@
+}
+
+template <typename T>
+void
+GroupIteratorsByType(
+    const std::vector<std::shared_ptr<knowhere::IndexNode::iterator>>&
+        iterators,
+    FieldId field_id,
+    int64_t topK,
+    Span<T> field_data,
+    std::vector<GroupByValueType>& group_by_values,
+    std::vector<int64_t>& seg_offsets,
+    std::vector<float>& distances,
+    const knowhere::MetricType& metrics_type) {
+    for (auto& iterator : iterators) {
+        GroupIteratorResult<T>(iterator,
+                               field_id,
+                               topK,
+                               field_data,
+                               group_by_values,
+                               seg_offsets,
+                               distances,
+                               metrics_type);
+    }
+}
+
+template <typename T>
+void
+GroupIteratorResult(
+    const std::shared_ptr<knowhere::IndexNode::iterator>& iterator,
+    FieldId field_id,
+    int64_t topK,
+    Span<T> field_data,
+    std::vector<GroupByValueType>& group_by_values,
+    std::vector<int64_t>& offsets,
+    std::vector<float>& distances,
+    const knowhere::MetricType& metrics_type) {
+    //1.
+    std::unordered_map<T, std::pair<int64_t, float>> groupMap;
+    std::vector<int64_t> tmpOffsets;
+    std::vector<float> tmpDistances;
+
+    //2. do iteration for at most 10 rounds
+    int round = 0;
+    while (round < 10) {
+        int64_t count = 0;
+        while (iterator->HasNext()) {
+            auto nextPair = iterator->Next();
+            int64_t offset = nextPair.first;
+            float dis = nextPair.second;
+            tmpOffsets.emplace_back(offset);
+            tmpDistances.emplace_back(dis);
+            count++;
+            if (count >= topK) {
+                break;
+            }
+        }
+        round++;
+        GroupOneRound<T>(tmpOffsets, tmpDistances, field_data, groupMap);
+        tmpOffsets.clear();
+        tmpDistances.clear();
+        if (!iterator->HasNext() || groupMap.size() == topK)
+            break;
+    }
+
+    std::vector<std::pair<T, std::pair<int64_t, float>>> sortedGroupVals(
+        groupMap.begin(), groupMap.end());
+    auto customComparator = [&](const auto& lhs, const auto& rhs) {
+        if (PositivelyRelated(metrics_type))
+            return lhs.second.second > rhs.second.second;
+        else
+            return lhs.second.second < rhs.second.second;
+    };
+    std::sort(sortedGroupVals.begin(), sortedGroupVals.end(), customComparator);
+
+    //3. save groupBy results
+    for (auto iter = sortedGroupVals.cbegin(); iter != sortedGroupVals.cend();
+         iter++) {
+        group_by_values.emplace_back(iter->first);
+        offsets.push_back(iter->second.first);
+        distances.push_back(iter->second.second);
+    }
+
+    //4. padding topK results, extra memory consumed will be removed when reducing
+    for (std::size_t idx = groupMap.size(); idx < topK; idx++) {
+        offsets.push_back(INVALID_SEG_OFFSET);
```

Comment:
could we just ignore these? the `INVALID_SEG_OFFSET` makes it impossible to optimize the retrieving with SIMD

Reply:
Currently, the reduce logic is designed towards n-query, meaning that results for multiple vectors is assembled together into a single vector. So we need to use this TAG to differentiate the 'domain' for each 'qi'. This is actually a placeholder for reduce logic